src/test/scala/com/microsoft/hyperspace/index/BucketUnionTest.scala

```diff
@@ -0,0 +1,126 @@
+      BucketUnion(
+        Seq(df1.queryExecution.optimizedPlan, df1_2.queryExecution.optimizedPlan),
+        BucketSpec(1, Seq(), Seq()))
+    }
+
+    BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+  }
+
+  test("BucketUnionStrategy test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val bucket = BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+
+    assert(BucketUnionStrategy(bucket).collect {
+      case BucketUnionExec(_, _) => true
+    }.length == 1)
+
+    assert(BucketUnionStrategy(df1.queryExecution.optimizedPlan).collect {
+      case BucketUnionExec(_, _) => true
+    }.isEmpty)
+  }
+
+  test("BucketUnionExec test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1 = df1.repartition(10)
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1_1 = df1_1.repartition(9)
+    val p1_2 = df1_1.repartition(10)
+
+    // different number of partition
+    intercept[AssertionError] {
+      val bucket = BucketUnion(
+        Seq(p1.queryExecution.optimizedPlan, p1_1.queryExecution.optimizedPlan),
+        BucketSpec(10, Seq(), Seq()))
+      spark.sessionState.executePlan(bucket).sparkPlan
+    }
+
+    val bucket = BucketUnion(
+      Seq(p1.queryExecution.optimizedPlan, p1_2.queryExecution.optimizedPlan),
+      BucketSpec(10, Seq(), Seq()))
+
+    assert(BucketUnionStrategy(bucket).collect {
+      case p: BucketUnionExec =>
+        assert(p.bucketSpec.numBuckets == 10)
+        assert(p.children.length == 2)
+        assert(p.output.length == p1.schema.fields.length)
+      true
+    }.length == 1)
+  }
+
+  test("BucketUnionRDD test") {
+    import spark.implicits._
+    val df1 = Seq((2, "name1"), (3, "name2")).toDF("id", "name")
+    val p1 = df1.repartition(10, $"id")
+    val df2 = Seq((2, "name1"), (3, "name2")).toDF("id", "name")
+    val p2 = df2.repartition(10, $"id")
+    val bucketSpec = BucketSpec(10, Seq("id"), Seq())
+
+    val rdd = new BucketUnionRDD[Row](spark.sparkContext, Seq(p1.rdd, p2.rdd), bucketSpec)
+    assert(rdd.getPartitions.length == 10)
+    assert(rdd.collect.length == 4)
+    assert(rdd.partitions.head.isInstanceOf[BucketUnionRDDPartition])
+
+    val partitionSum = rdd
+      .mapPartitionsWithIndex {
+        case (partitionNum, it) => Iterator.single(partitionNum -> it.map(r => r.getInt(0)).sum)
+      }
+      .collect()
+      .toSeq
+
+    // since rows with id=2 assigned to the same partition and likewise rows with id=3,
+    // summation of each partition should be one of [0, 4, 6, 10]
+    val availableSum = Set(0, 4, 6, 10)
+    assert(partitionSum.forall(p => availableSum.contains(p._2)))
```

Comment:
nit: we can avoid partition id:
```suggestion
    val partitionSum: Seq[Int] = rdd
      .mapPartitions(it => Iterator.single(it.map(r => r.getInt(0)).sum))
      .collect()
      .toSeq

    // since rows with id=2 assigned to the same partition and likewise rows with id=3,
    // summation of each partition should be one of [0, 4, 6, 10]
    val availableSum = Set(0, 4, 6, 10)
    assert(partitionSum.forall(p => availableSum.contains(p)))
```